http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/RuntimeServiceDbQueries.scala

```diff
@@ -264,6 +283,250 @@ object RuntimeServiceDbQueries {
+    )
+  }
+
+  /**
+   * Query the runtimes (CLUSTER) table with the given filters. Only return authorized resources (per reader*Ids and/or owner*Ids).
+   * @param labelMap
+   * @param excludeStatuses
+   * @param creatorOnly
+   * @param workspaceId
+   * @param cloudContextOrCloudProvider
+   * @param readerRuntimeIds
+   * @param readerWorkspaceIds
+   * @param ownerWorkspaceIds
+   * @param readerGoogleProjectIds
+   * @param ownerGoogleProjectIds
+   * @return
+   */
+  private def listAuthorizedRuntimesHelper(
+    // Filters
+    labelMap: LabelMap,
+    excludeStatuses: List[RuntimeStatus],
+    creatorOnly: Option[WorkbenchEmail],
+    workspaceId: Option[WorkspaceId],
+    cloudContextOrCloudProvider: Option[Either[CloudContext, CloudProvider]],
+
+    // Authorizations
+    readerRuntimeIds: List[String],
+    readerWorkspaceIds: List[String],
+    ownerWorkspaceIds: List[String],
+    readerGoogleProjectIds: List[String],
+    ownerGoogleProjectIds: List[String]
+  ): DBIO[Vector[ListRuntimeResponse2]] = {
+
+    // Authorize: show only resources the user is permitted to see
+    val filterVisibleIds = getFilterVisibleIds(readerRuntimeIds,
+                                               readerWorkspaceIds,
+                                               ownerWorkspaceIds,
+                                               readerGoogleProjectIds,
+                                               ownerGoogleProjectIds
+    )
+
+    // Filter: show only user selections
+    val filterWorkspaceId = workspaceId match {
+      case Some(wid) =>
+        List(s"C.`workspaceId` = '${wid.value.toString}'")
+      case None => List.empty
+    }
+
+    val filterCloud = cloudContextOrCloudProvider match {
+      case Some(Left(CloudContext.Gcp(gp))) =>
+        List(s"C.`cloudProvider` = ${CloudProvider.Gcp} AND C.`cloudContext` = '${gp.value}'")
+      case Some(Left(CloudContext.Azure(actx))) =>
+        List(s"C.`cloudProvider` = ${CloudProvider.Azure} AND C.`cloudContext` = '${actx.asString}'")
+      case Some(Right(cloudProvider)) => List(s"C.`cloudProvider` = '${cloudProvider.asString}'")
+      case None                       => List.empty
+    }
+
+    val filterNotExcludedStatus = excludeStatuses.map(s => s"C.`status` != '${s.toString}'")
+
+    val filterCreator = creatorOnly match {
+      case Some(creator) => List(s"C.`creator` = '${creator.value}'")
+      case None          => List.empty
+    }
+
+    val filterClusters = (
+      filterVisibleIds ++
+        filterWorkspaceId ++
+        filterCloud ++
+        filterNotExcludedStatus ++ filterCreator
+    ).filterNot(_.isEmpty).mkString(") AND (")
+
+    val whereFilterClusters = if (filterClusters.isEmpty) "" else s"where (${filterClusters})"
+
+    val whereFilterLabels =
+      if (labelMap.isEmpty)
+        ""
+      else {
+        val query = labelMap
+          .map { case (k, v) =>
+            s"(LABEL.key = '${k}' and LABEL.value = '${v}')"
```

Comment:
I haven't fully processed this, but am a little uncomfortable with raw string interpolation in SQL fragments with user-specified values (the label query in this case). Will 🤔 more…

Reply:
OK I see the previous code did similar.. which I also don't like 😅 . But keeping that in mind..

Reply:
Agreed that this is bad practice and insecure! I had a stab at converting to Slick but I don't know the syntax well enough to be confident. I had to leave this refactor for another time (https://broadworkbench.atlassian.net/browse/IA-4626)